### PR TITLE
Show payment methods in vertical saved list

### PR DESCRIPTION
**Pull request by porter-stripe**
## Summary

- Adds ability to display and select payment methods in the `VerticalSavedPaymentOptionsViewController`
- Adds delegates to communicate between the new `PaymentMethodRowButton` <> `VerticalSavedPaymentOptionsViewController` <> `PaymentSheetVerticalViewController `
- Still grappling with the janky push animation, will fix later.

https://github.com/stripe/stripe-ios/assets/88012362/b3863e51-cd68-4205-ad11-578cd05f88c1

## Motivation
https://jira.corp.stripe.com/browse/MOBILESDK-2024

## Testing
- Manual
- Snapshots

## Changelog
N/A

### CODE REVIEW

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/CheckmarkCircleView.swift [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+//  StripePaymentSheet
+//
+//  Created by Nick Porter on 5/9/24.
+//
+
+import Foundation
+import UIKit
+
+/// Draws a circle with the desired fill color with a white checkmark in the center
+final class CheckmarkCircleView: UIView {
```
Comment (porter-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/CheckmarkCircleView.swift:12]: Omitted snapshot tests for this class and relying on the `VerticalSavedPaymentOptionsViewControllerSnapshotTests`. Can add them if we feel they will be useful.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/PaymentMethodRowButton.swift [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+@_spi(STP) import StripePaymentsUI
+@_spi(STP) import StripeUICore
+import UIKit
+
+protocol PaymentMethodRowDelegate: AnyObject {
+    func didSelectButton(_ button: PaymentMethodRowButton)
+    // TODO(porter) Add did delete and did update
+}
+
+final class PaymentMethodRowButton: UIView {
```
Comment (porter-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/PaymentMethodRowButton.swift:19]: Same, omitted snapshot tests for this view and relied on `VerticalSavedPaymentOptionsViewControllerSnapshotTests`.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/VerticalSavedPaymentOptionsViewController.swift [hunk 3]
```diff
@@ -27,21 +37,33 @@ class VerticalSavedPaymentOptionsViewController: UIViewController {
 
     private lazy var headerLabel: UILabel = {
         let label = PaymentSheetUI.makeHeaderLabel(appearance: configuration.appearance)
-        label.text = .Localized.select_payment_method
+        let nonCardPaymentMethods = paymentMethods.filter({ $0.type != .card })
+        label.text = nonCardPaymentMethods.isEmpty ? .Localized.select_card : .Localized.select_payment_method
         return label
     }()
+    
+    private lazy var paymentMethodRows: [(paymentMethod: STPPaymentMethod, button: PaymentMethodRowButton)] = {
```
Comment (porter-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/VerticalSavedPaymentOptionsViewController.swift:45]: Not sure how I feel about this. It's tricky to associate a `STPPaymentMethod` with a row button when:
1. Not using a table view
2. Not muddying up the `PaymentMethodRowButton.ViewModel` with an `STPPaymentMethod`.

Open to suggestions

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/VerticalSavedPaymentOptionsViewController.swift [hunk 4]
```diff
@@ -1,91 +0,0 @@
-//
-//  VerticalSavedPaymentOptionsViewController.swift
```
Comment (porter-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/VerticalSavedPaymentOptionsViewController.swift:2]: ooof, git doesn't recognize it was moved :(

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/PaymentMethodRowButton.swift [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    init(viewModel: ViewModel) {
+        self.viewModel = viewModel
+        super.init(frame: .zero)
+
+        addSubview(shadowRoundedRect)
+        NSLayoutConstraint.activate([
+            shadowRoundedRect.topAnchor.constraint(equalTo: topAnchor),
+            shadowRoundedRect.leadingAnchor.constraint(equalTo: leadingAnchor),
+            shadowRoundedRect.trailingAnchor.constraint(equalTo: trailingAnchor),
+            shadowRoundedRect.bottomAnchor.constraint(equalTo: bottomAnchor),
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/PaymentMethodRowButton.swift:99]: Delete and replace with `addAndPinSubview()`?

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/PaymentMethodRowButton.swift [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            shadowRoundedRect.heightAnchor.constraint(equalToConstant: height),
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/PaymentMethodRowButton.swift:100]: Why is this needed?

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/VerticalSavedPaymentOptionsViewController.swift (hunk 3 above)
Comment (yuki-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/VerticalSavedPaymentOptionsViewController.swift:44]: I don't have strong ideas either. Why not have the payment method as a property of the view (either in the viewmodel or just directly)?

File: StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/VerticalSavedPaymentOptionsViewControllerSnapshotTests.swift [hunk 7]
```diff
@@ -37,4 +37,12 @@ final class VerticalSavedPaymentOptionsViewControllerSnapshotTests: STPSnapshotT
         sut.view.autosizeHeight(width: 375)
         STPSnapshotVerifyView(sut.view)
     }
+
+    private func generatePaymentMethods() -> [STPPaymentMethod] {
+        return [STPFixtures.paymentMethod(),
+                STPFixtures.usBankAccountPaymentMethod(),
+                STPFixtures.usBankAccountPaymentMethod(bankName: "BANK OF AMERICA"),
+                STPFixtures.usBankAccountPaymentMethod(bankName: "STRIPE"),
+                STPFixtures.sepaDebitPaymentMethod(), ]
```
Comment (yuki-stripe) [inline StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/VerticalSavedPaymentOptionsViewControllerSnapshotTests.swift:46]: We have two ways to provide fixtures unfortunately.  THe old Obj-C way is static methods on `STPFixtures` class. THe new way is to provide extensions on the thing. See `STPFixtures+PaymentSheet.swift`. It already has `STPPaymentMethod._testSEPA` and I think it's a more ergonomic and scalable way we should use going forward.

Comment (yuki-stripe) [reply StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/VerticalSavedPaymentOptionsViewControllerSnapshotTests.swift:46]: Should probably make a ticket to migrate us off of `STPFixtures`

**Review by yuki-stripe — COMMENTED** [review]
Awesome!

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentOption+Images.swift [hunk 8]
```diff
@@ -97,6 +97,24 @@ extension STPPaymentMethod {
             return makeIcon()
         }
     }
+
+    /// Returns an image to display inside a row representing the given payment option in the saved PM row view
+    func makeSavedPaymentMethodRowImage() -> UIImage {
```
[the remark this answers is not in the document]
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentOption+Images.swift:102]: `makeSavedPaymentMethodCellImage` uses carousel images for card brands and for SEPA. We need the images that are used in the PAN field for card brand rather than the carousel images. Here the diff if we use `makeSavedPaymentMethodCellImage`, notice visa & SEPA look a little different than what is laid out in the mocks. The differences would be more sark in dark mode but I don't think we have mocks for dark mode.
CleanShot 2024-05-09 at 17 57 13


#### Mock
CleanShot 2024-05-09 at 17 58 11

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/PaymentMethodRowButton.swift [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+final class PaymentMethodRowButton: UIView {
+
+    struct ViewModel {
+        let appearance: PaymentSheet.Appearance
+        let text: String
+        let image: UIImage
+        // TODO(porter) Add can remove and can update
+    }
```
[the remark this answers is not in the document]
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/PaymentMethodRowButton.swift:26]: Probably not totally needed right now, but I anticipate this will scale better and am hoping in the future to create view models for `PaymentMethodRowButton` from a `STPPaymentMethod`. Mainly trying to keep the size of `PaymentMethodRowButton.init` down and offload the injection building somewhere else. Not totally needed but kinda like the direction.

```swift
extension STPPaymentMethod {
   var rowButtonViewModel: PaymentMethodRowButton.ViewModel {...}
}

// Somewhere else
let rowButtons = savedPaymentMethods.map{PaymentMethodRowButton(appearance: appearance, $0.rowButtonViewModel)}
```

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/PaymentMethodRowButton.swift (hunk 6 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/PaymentMethodRowButton.swift:100]: Without explicitly setting the height of the `shadowRect` it is just sized the minimum to fit the content. Is there a better way to do this?

Removing the `shadowRoundedRect.heightAnchor.constraint(equalToConstant: height)`:
CleanShot 2024-05-09 at 18 11 17

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/VerticalSavedPaymentOptionsViewController.swift (hunk 3 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/VerticalSavedPaymentOptionsViewController.swift:44]: I just felt like the `PaymentMethodRowButton` and `PaymentMethodRowButton.ViewModel` shouldn't know about an entire `STPPaymentMethod` and it should just know the data it needs to draw itself, no other reason.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/PaymentMethodRowButton.swift (hunk 6 above)
Comment (yuki-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/PaymentMethodRowButton.swift:96]: Looks like the stack view doesn't have top or bottom margins.

```
stackView.directionalLayoutMargins = PaymentSheetUI.defaultMargins
```

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentOption+Images.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (yuki-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentOption+Images.swift:102]: Ohh right, we have 'carousel' versions of these. I guess we can delete those once we finish LPM visibility and the horizontal saved pm carousel goes away.

[the remark this answers is not in the document]
Comment (yuki-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentOption+Images.swift:102]: Ohh right, we have 'carousel' versions of these. I guess we can delete those once we finish LPM visibility and the horizontal saved pm carousel goes away.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/PaymentMethodRowButton.swift (hunk 6 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Saved Payment Method Screen/Vertical Saved Payment Method Screen/PaymentMethodRowButton.swift:96]: Oh good catch, ty!

File: StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/VerticalSavedPaymentOptionsViewControllerSnapshotTests.swift (hunk 7 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/VerticalSavedPaymentOptionsViewControllerSnapshotTests.swift]: Filed [here!](https://jira.corp.stripe.com/browse/RUN_MOBILESDK-3166). I'll leave as-is for now but I'll pick this up next time I'm on run.
